File: python/partition-array-for-maximum-sum.py
class Solution:
    def maxSumAfterPartitioning(self, arr: List[int], k: int) -> int:
        n = len(arr)
        memory = [0] * (n + 1)
        for i in range(n):
            current_maximal = arr[i]
            for j in range(k):
                l = i - j
                if l < 0:
                    break
                current_maximal = max(current_maximal, arr[l])
                candidate = memory[l] + current_maximal * (j + 1)
                memory[i + 1] = max(memory[i + 1], candidate)
        return memory[n]

File: go/partition-array-for-maximum-sum.go
func maxSumAfterPartitioning(arr []int, k int) int {
    n := len(arr)
    memory := make([]int, n + 1)
    for i := 0; i < n; i++ {
        current_maximal := arr[i]
        for j := 0; j < k; j++ {
            l := i - j
            if l < 0 {
                break
            }
            if current_maximal < arr[l] {
                current_maximal = arr[l]
            }
            candidate := memory[l] + (current_maximal * (j + 1))
            if memory[i + 1] < candidate {
                memory[i + 1] = candidate
            }
        }
    }
    return memory[n]
}
